### PR TITLE
Fix(Normalization): Denormalize with target stats when present

### LINKED ISSUE #967: Normalization: target stats not used for denomalization
## Describe the bug

The denormalization method is used to transform the distribution of the model's output to the distribution of the target data. However in both the `MeanStdNormalization` and `RangeNormalization` the input stats are used. For methods that have a different number of input and output channels this will result in an error.

https://github.com/CAREamics/careamics/blob/a9033f51acff3c60a8725cc35fce977791160019/src/careamics/dataset/normalization/mean_std_normalization.py#L176-L179

## Suggestion

1. Have two methods `denormalize_input` and `denormalize_target`.

OR

2. If targets don't exist, use input stats otherwise use target stats for denormalization.

### PR DESCRIPTION
**Pull request by parallelArchitect**
## Disclaimer

- [ ] I am an AI agent.
- [X] I have used AI and I thoroughly reviewed every line.
- [ ] I have not used AI extensively.

## Description

> [!NOTE]  
> **tldr**: denormalize() now correctly uses target statistics instead of input statistics when both are provided.

### Background - why do we need this PR?

Both denormalize() methods unconditionally used input statistics to reverse-normalize the model's output. The model predicts values in the target distribution, not the input distribution. When input and target have different channel counts, this causes a ValueError in broadcast_stats. When channel counts match, the bug is silent but produces wrong values — the output is scaled to the input range instead of the target range.

### Overview - what changed?

denormalize() in both classes now checks for target stats first and falls back to input stats only when none were provided, preserving existing behavior.

### Implementation - how did you implement the changes?

Added a conditional check at the top of each denormalize() method — if target_means/target_stds (or target_mins/target_maxes for RangeNormalization) are not None, use them; otherwise fall back to input stats. No changes to the normalization forward pass or any other methods. 

## Changes Made

 denormalize() in src/careamics/dataset/normalization/mean_std_normalization.py
denormalize() in src/careamics/dataset/normalization/range_normalization.py

### New features or files
- test_denormalize_uses_target_stats_not_input_stats in tests/dataset/normalization/test_mean_std_normalization.py
test_denormalize_uses_target_range_not_input_range in tests/dataset/normalization/test_minmax_normalization.py
...

## How has this been tested?

Both regression tests reproduce the exact failure case from the issue (mismatched input/target channel counts), confirm the fix produces correct values, and verify backward compatibility when no target stats are provided. All 63 normalization tests pass locally.

## Related Issues

- Resolves #967

## Breaking changes

None — existing behavior preserved when no target stats are provided. 

---

**Please ensure your PR meets the following requirements:**

- [X]  Code builds and passes tests locally, including doctests
- [X] New tests have been added (for bug fixes/features)
- [ ] Documentation has been updated
- [X] Pre-commit passes

### CODE REVIEW

File: src/careamics/dataset/normalization/mean_std_normalization.py [hunk 1]
```diff
@@ -157,6 +157,11 @@ def _apply_denormalization(
     def denormalize(self, patch: torch.Tensor) -> torch.Tensor:
         """Reverse the normalization operation for a batch of patches.
 
+        Uses target statistics when available, since the patch passed in
+        is the model's output (in the target's distribution), not the
+        original input. Falls back to input statistics when no target
+        statistics were provided, preserving prior behavior.
```
Comment (melisande-c) [inline src/careamics/dataset/normalization/mean_std_normalization.py]: Do not reference "prior behaviour" in documentation since this will not make sense to people reading it later, especially in the API reference on the website.

The algorithm N2V is self-supervised and the input statistics are the target statistics which is why we want this behaviour.

```suggestion
    The data is denormalized using the target statistics, when available. When no target
    statistics are available, which is the case for self-supervised algorithms such as
    N2V, the data is denormalized using the input statistics.
```

File: src/careamics/dataset/normalization/range_normalization.py [hunk 2]
```diff
@@ -113,6 +102,11 @@ def __call__(
     def denormalize(self, patch: torch.Tensor) -> torch.Tensor:
         """Reverse the normalization operation for a batch of patches.
 
+        Uses target range when available, since the patch passed in is the
+        model's output (in the target's distribution), not the original input.
+        Falls back to input range when no target range was provided, preserving
+        prior behavior.
```
Comment (melisande-c) [inline src/careamics/dataset/normalization/range_normalization.py]: Again, do not mention "prior behaviour" in documentation. Replace this segment with something similar suggested in my previous [comment](https://github.com/CAREamics/careamics/pull/977/changes#r3498437862)

File: tests/dataset/normalization/test_mean_std_normalization.py [hunk 3]
```diff
@@ -226,3 +226,48 @@ def test_global_stats_pools_across_channels():
...
     assert sample.data[0].mean() < 0.0
     assert sample.data[1].mean() > 0.0
+
+
+def test_denormalize_uses_target_stats_not_input_stats():
+    """Regression test for #967: denormalize() must use target stats when
+    they exist, not input stats. Real failure case: input and target have
+    a different number of channels (e.g. denoising task), so reusing
+    input's per-channel stats on the model's (target-shaped) output either
+    crashes or silently applies the wrong values."""
```
Comment (melisande-c) [inline tests/dataset/normalization/test_mean_std_normalization.py]: Do not mention the issue number or PR number in the test docstring or any of the comments. Let's not call this a regression test, we are just testing new behaviour, so let's rename it to something like `test_denormalize_with_target_stats`.

File: tests/dataset/normalization/test_mean_std_normalization.py [hunk 4]
```diff
@@ -226,3 +226,48 @@ def test_global_stats_pools_across_channels():
...
+    )
+    dataset = create_dataset(config=config, inputs=[input_data], targets=[target_data])
+
+    # Real failure scenario from the issue: model output has a different
+    # channel count than the input. Force target stats to be multi-channel
+    # while input stats stay single-channel (global), the exact mismatch
+    # described in #967.
+    norm = dataset.normalization
+    norm.target_means = [150.0, 300.0]
+    norm.target_stds = [30.0, 50.0]
```
Comment (melisande-c) [inline tests/dataset/normalization/test_mean_std_normalization.py]: set the target stats when creating the configuration above, not here.

File: tests/dataset/normalization/test_mean_std_normalization.py [hunk 5]
```diff
@@ -226,3 +226,48 @@ def test_global_stats_pools_across_channels():
...
+
+    # Simulate model output: batch=1, 2 channels (matches target, not input)
+    model_output = torch.zeros((1, 2, 8, 8))
```
Comment (melisande-c) [inline tests/dataset/normalization/test_mean_std_normalization.py]: can you make the model output `torch.ones`? So that the result should be `target_std` + `target_mean` not just the `target_mean`

File: tests/dataset/normalization/test_minmax_normalization.py [hunk 6]
```diff
@@ -221,3 +221,40 @@ def test_global_stats_pools_across_channels():
     sample, *_ = dataset[0]
     assert np.isclose(sample.data[0].mean(), 0.0, atol=0.01)
     assert np.isclose(sample.data[1].mean(), 1.0, atol=0.01)
+
+
+def test_denormalize_uses_target_range_not_input_range():
+    """Regression test for #967: denormalize() must use target range when
+    it exists, not input range. Real failure case: input and target have
+    a different number of channels, so reusing input's per-channel range
+    on the model's (target-shaped) output gives silently wrong values."""
```
Comment (melisande-c) [inline tests/dataset/normalization/test_minmax_normalization.py]: Again, let's make this docstring a simple one-liner to describe what the test does without mentioning the PR number or the issue number, and rename the test to `test_denormalize_with_target_range`.

File: tests/dataset/normalization/test_minmax_normalization.py [hunk 7]
```diff
@@ -221,3 +221,40 @@ def test_global_stats_pools_across_channels():
...
+            "input_maxes": [100.0],
+            "target_mins": [0.0],
+            "target_maxes": [1000.0],
+        },
+    )
+    dataset = create_dataset(config=config, inputs=[input_data], targets=[target_data])
+
+    norm = dataset.normalization
+    norm.target_mins = [0.0, 0.0]
+    norm.target_maxes = [1000.0, 2000.0]
```
Comment (melisande-c) [inline tests/dataset/normalization/test_minmax_normalization.py]: Again, set the target stats when creating the configuration instead of patching it in here.

**Review by melisande-c — CHANGES_REQUESTED** [review]
Hi @parallelArchitect, thank you for your contribution, however next time please read our contribution guidelines which states that people outside the core team who wish to contribute should get in contact with us first, you can request to work on an existing issue by commenting on it and waiting for our response on how to proceed. 

I have requested some changes, mostly about documentation and the tests.

Thanks again!

**Comment by parallelArchitect** [comment]
Thanks for the thorough review @melisande-c — addressed everything:

- Removed "prior behavior" language from both docstrings, used your suggested phrasing
- Renamed tests to `test_denormalize_with_target_stats` / `test_denormalize_with_target_range`, simplified docstrings, removed issue/PR references
- Moved target stats/range into config creation instead of patching after dataset creation
- Changed mean_std test's model output from zeros to ones so the assertion validates target_mean + target_std being applied, not just the mean

All normalization tests pass (63/63).

**Comment by melisande-c** [comment]
Great thanks for the changes, can you fix the last pre-commit "line too long" error and then we are good to merge.

**Review by melisande-c — APPROVED** [review]
